engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java

```diff
@@ -0,0 +1,113 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long end;
+
+  public ArraySlice(Object storage, long start, long end) {
+    if (storage instanceof ArraySlice slice) {
+      this.storage = slice.storage;
+      this.start = slice.start + start;
+      this.end = Math.min(slice.end, slice.start + end);
+    } else {
+      if (CompilerDirectives.inInterpreter()) {
+        if (!InteropLibrary.getUncached().hasArrayElements(storage)) {
+          throw new IllegalStateException("Vector needs array-like delegate, but got: " + storage);
+        }
+      }
+
+      this.storage = storage;
+      this.start = start;
+      this.end = end;
+    }
+  }
+
+  /**
+   * Marks the object as array-like for Polyglot APIs.
+   *
+   * @return {@code true}
+   */
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  public long getArraySize(@CachedLibrary(limit = "3") InteropLibrary interop)
+      throws UnsupportedMessageException {
+    long storageSize = interop.getArraySize(storage);
+    return Math.min(storageSize, end) - start;
```

Comment:
This could get negative, right?

Creates slice over JavaScript array and then remove something from the underlying array. The `storageSize` gets smaller and substracting `start` may give us negative index.